ops/pebble.py

```diff
@@ -1973,6 +1987,101 @@ def read1(self, n: int = -1) -> Union[str, bytes]:
         return self.read(n)
 
 
+class IdentityAccess(str, enum.Enum):
+    """Enum of identity access levels."""
+
+    ADMIN = 'admin'
+    READ = 'read'
+    METRICS = 'metrics'
+    UNTRUSTED = 'untrusted'
+
+
+@dataclasses.dataclass
+class LocalIdentity:
+    """Local identity configuration (for Unix socket / UID authentication)."""
+
+    user_id: int
+
+    @classmethod
+    def from_dict(cls, d: LocalIdentityDict) -> LocalIdentity:
+        """Create new LocalIdentity from dict parsed from JSON."""
+        return cls(user_id=d['user-id'])
+
+    def to_dict(self) -> LocalIdentityDict:
+        """Convert this local identity to its dict representation."""
+        return {'user-id': self.user_id}
+
+
+@dataclasses.dataclass
+class BasicIdentity:
+    """Basic identity configuration (for HTTP basic authentication)."""
+
+    password: str
+    """sha512-crypt-hashed password.
+
+    When writing, this is a SHA512-crypt hashed password. When reading, it is
+    returned as ``*****``.
+
+    Use ``openssl passwd -6`` to generate a hashed password (sha512-crypt format).
+    """
+
+    @classmethod
+    def from_dict(cls, d: BasicIdentityDict) -> BasicIdentity:
+        """Create new BasicIdentity from dict parsed from JSON."""
+        return cls(password=d['password'])
+
+    def to_dict(self) -> BasicIdentityDict:
+        """Convert this basic identity to its dict representation."""
+        return {'password': self.password}
+
+
+@dataclasses.dataclass
+class Identity:
+    """Pebble identity configuration."""
+
+    access: Union[IdentityAccess, str]
+    local: Optional[LocalIdentity] = None
+    basic: Optional[BasicIdentity] = None
+
+    def __post_init__(self) -> None:
+        """Validate that at least one of local or basic is provided."""
+        if self.local is None and self.basic is None:
+            raise ValueError('at least one of "local" or "basic" must be provided')
+
+    @classmethod
+    def from_dict(cls, d: IdentityDict) -> Identity:
+        """Create new Identity from dict parsed from JSON."""
+        if 'access' not in d:
+            raise KeyError('"access" key is required in IdentityDict')
```

Comment:
I think we can drop this again, as `KeyError` will be raised by the `d['access']` expression below.

Reply:
These two changes are to make pyright happy.

We could not drop `if 'access' not in d`, because `access` is not a required key in `IdentityDict` ([reportTypedDictNotRequiredAccess](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportTypedDictNotRequiredAccess)), also see [here](https://github.com/canonical/operator/pull/1672#discussion_r2036576287).

The reason for ` access: Union[IdentityAccess, str]` is that access might be a string, so pyright would complain when we do `return cls(access=access, ...)`.

Reply:
Okay, yep, fine to leave the `raise KeyError` then.

On the type annotation -- yes, that's expected, but we want to simplify the typing for the user and pretend. So I think we want to silent Pyright and do this:

```python
try:
    access = IdentityAccess(d['access'])
except ValueError:
    access = typing.cast(IdentityAccess, d['access'])
```